File: 07_linked_list/07_09_delete_middle_node_of_linked_list/delete_middle_node_of_linked_list.cpp
#include <iostream>

class ListNode {
public:
    int val;
    ListNode* next;

    ListNode(int value = 0, ListNode* nextNode = nullptr) : val(value), next(nextNode) {}
};

ListNode* deleteMiddle(ListNode* head) {
    if (!head || !head->next) {
        return nullptr;
    }

    ListNode* slow = head;
    ListNode* fast = head->next->next;

    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
    }

    slow->next = slow->next->next;
    return head;
}

// Helper function to print the linked list
void printList(const std::string& label, ListNode* head) {
    std::cout << label << " ";
    ListNode* current = head;
    while (current) {
        std::cout << current->val << " ";
        current = current->next;
    }
    std::cout << std::endl;
}

int main() {
    // Test Case 1: Odd number of nodes
    ListNode* head1 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
    printList("Test Case 1 (Before):", head1);
    ListNode* result1 = deleteMiddle(head1);
    printList("Test Case 1 (After):", result1);  // Output: 1 -> 2 -> 4 -> 5

    // Test Case 2: Even number of nodes
    ListNode* head2 = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4))));
    printList("Test Case 2 (Before):", head2);
    ListNode* result2 = deleteMiddle(head2);
    printList("Test Case 2 (After):", result2);  // Output: 1 -> 2 -> 4

    // Test Case 3: Single node
    ListNode* head3 = new ListNode(1);
    printList("Test Case 3 (Before):", head3);
    ListNode* result3 = deleteMiddle(head3);
    printList("Test Case 3 (After):", result3);  // Output: nullptr

    // Remember to free allocated memory
    delete head1;
    delete head2;
    delete head3;

    return 0;
}

File: 07_linked_list/07_09_delete_middle_node_of_linked_list/delete_middle_node_of_linked_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def delete_middle(head):
    if not head or not head.next:
        return None

    slow = head
    fast = head.next.next

    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next

    slow.next = slow.next.next
    return head

# Helper function to print the linked list
def print_list(label, head):
    print(f"{label} ", end="")
    current = head
    while current:
        print(current.val, end=" ")
        current = current.next
    print()

# Test Case 1: Odd number of nodes
head1 = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
print_list("Test Case 1 (Before):", head1)
result1 = delete_middle(head1)
print_list("Test Case 1 (After):", result1)  # Output: 1 -> 2 -> 4 -> 5

# Test Case 2: Even number of nodes
head2 = ListNode(1, ListNode(2, ListNode(3, ListNode(4))))
print_list("Test Case 2 (Before):", head2)
result2 = delete_middle(head2)
print_list("Test Case 2 (After):", result2)  # Output: 1 -> 2 -> 4

# Test Case 3: Single node
head3 = ListNode(1)
print_list("Test Case 3 (Before):", head3)
result3 = delete_middle(head3)
print_list("Test Case 3 (After):", result3)  # Output: None
